### ISSUE TITLE
Service Bus methods do not work properly on a partitioned queue

### ISSUE THREAD

**Issue by jtmilne** [Azure/azure-sdk-for-node#1415]
The standard send and receive methods do not work properly against a partitioned queue (the default option when creating a new queue in the portal). Issues encountered include: 

-extremely slow response times
-duplicate message delivery
-loss of FIFO

See [my StackOverflow post here](http://stackoverflow.com/questions/30089590/polling-an-azure-service-bus-queue-from-an-azure-webjob-using-node-js/) for detailed description and log outputs.

**Comment by Mobius5150** [comment]
I have been using service bus for a while now and recently noticed that my consumers had recently started operating nowhere near even 1/100th of their previous performance. It looks like I'm experiencing the same issue as @jtmilne. I was however able to go back to version 0.10.2 (December 2014) of the Azure node SDK and everything appears to work nice and fast again. I haven't tried any more recent versions.

**Comment by denmerc** [comment]
the fact

**Comment by djrosanova** [comment]
About the latency - Service Bus Standard and Basic are a shared platform - so there can be times when performance changes. We did recently have some capacity crunch due to traffic growth - but those should be resolved. The SDK for Node hasn't changed much in quite a while so I don't think duplicate messages and FIFO were introduced in or changed. Speaking of FIFO - that's making a lot of assumptions and really only works with HTTP (i.e. Node) if you're only using one consumer and it reads in order - which may or may not be how you've implemented it. Calling receiveQueueMessage in Node is an async parallel operation like everything else in node. I was just testing today and getting performance as expected. Are you long polling?

**Comment by jtmilne** [comment]
@djrosanova All the code used and results is provided in the link to the StackOverflow post.

**Comment by djrosanova** [comment]
So I looked at the code - thanks for posting. The issue is definitely that it's not clear what a partitioned entity means or what long polling means in that context. I'm also not sure how you really plan to use in order messaging with node. Seems like you'd bypass the async nature that makes node so useful in the first place.

Here's what a partitioned entity means: use a lot of shards to improve availability / performance. When you do this calling receive with a long poll has to check across all partitions to see where a message is - via HTTP this is sub-optimal. Also as you notied - partition is default on portal, but not in the Node SDK - so if you create your queue in code (a good practice anyway normally) then you won't have this issue.

You could also try checking out this set of node packages for using AMQP with Service Bus - which is WAY better performance. https://github.com/noodlefrenzy 

BTW: here's what I got from your code:
listener job started
received a response in 0s seconds
received test message number 1
received a response in 16s seconds
received test message number 5
received a response in 2s seconds
received test message number 2
received a response in 17s seconds
received test message number 9
received a response in 0s seconds
received test message number 6
received a response in 0s seconds
received test message number 3
received a response in 12s seconds
received test message number 8
received a response in 22s seconds
received test message number 0
received a response in 26s seconds
received test message number 7
received a response in 20s seconds
received test message number 4
received a response in 30s seconds
received test message number 10

**Comment by jtmilne** [comment]
@djrosanova I really don't understand your explanation. 

You state above that partitioned entities "improve performance". And your documentation clearly states that queues deliver messages in FIFO (https://azure.microsoft.com/en-us/documentation/articles/service-bus-nodejs-how-to-use-queues/).

However, from your test above it takes 2:25 to receive 10 messages and they are not in FIFO order for a partitioned queue. From my tests they are *never* FIFO when a queue is partitioned. When running the same test on a non-partitioned entity it takes ~1s and they are always FIFO. 

If you feel that a ~150x degradation in performance is acceptable and not an actual bug, and that random order delivery is acceptable and not a bug, then your documentation should clearly state that so people don't waste days figuring out why it's not working as advertised in the docs. The reality is that this library is not functional with partitioned queues and the docs should reflect that.
